### PR TITLE
Expose additional information (expiration, sha1 hash) about realm keys in admin UI

### LINKED ISSUE #17743: Show expiration date of realm keys
### Description

The realm keys are currently listed like this:
```
Algorithm  | Type | Kid                                         | Provider          | Public keys
RSA-OAEP   | RSA  | 7ShgKSatq0nK5fyXxvj-fwrXj28AZ68FzHTK2memURQ | rsa-enc-generated | [Public Key] [Certificate]
RS256      | RSA  | WvCsvncC4MtdlyCmndluWsbaAoK7MBMORiD4Ai25YyU | rsa-generated     | [Public Key] [Certificate]
```

It would be helpful to also provide "valid from" and "valid to" information and the thumbprint value of the certificate.
Valid to should be displayed in the table:
```
Algorithm  | Type | Kid                                         | Provider          | Valid to            | Public keys
RS256      | RSA  | WvCsvncC4MtdlyCmndluWsbaAoK7MBMORiD4Ai25YyU | rsa-generated     | 2030-11-27T17:42:02 | [Public Key] [Certificate]
```
Other details could be shown when clicking on the [Certificate] button.

Current workaround for the missing feature:  

- Click [Certificate] button
- Copy the displayed key into a file `cert.pem`
- Add line -----BEGIN CERTIFICATE----- as first line and -----END CERTIFICATE----- as last line
- Use openssl to parse the certificate: `openssl x509 -in cert.pem -text`

### Discussion

PR which also adds this feature at old admin console as a side effect:
[KEYCLOAK-17609: OIDC broker x5t jws header #7890](https://github.com/keycloak/keycloak/pull/7890)

Realm key related discussions
[Automatic realm key rotation #13776](https://github.com/keycloak/keycloak/discussions/13776)
[Realm key(s) uses - Allowing multivalue fields #8523](https://github.com/keycloak/keycloak/discussions/8523)

### Motivation

It is important to see when a key will expire.

### Details

Keycloak Admin API Backend needs to provide additional data like _from_, _to_, _thumbprint_ with the /keys resource.

### PR DESCRIPTION
**Pull request by thomasdarimont**
- Show Not After information
- Show certificate sha1 thumbprint

Incorporates: #7890

Fixes #17743 and KEYCLOAK-17609

### CODE REVIEW

**Comment by thomasdarimont** [comment]
Just tested this again locally, looks like that with current main:

image

File: js/apps/admin-ui/src/realm-settings/keys/KeysListTab.tsx [hunk 1]
```diff
@@ -221,6 +240,17 @@ export const KeysListTab = ({ realmComponents }: KeysListTabProps) => {
                     >
                       {t("certificate")}
                     </Button>
+                    <Button
```
Comment (thomasdarimont) [inline js/apps/admin-ui/src/realm-settings/keys/KeysListTab.tsx:243]: @edewit  The button alignment seems a bit off, how can I fix that?

File: js/apps/admin-ui/src/realm-settings/keys/KeysListTab.tsx [hunk 2]
```diff
@@ -180,10 +192,23 @@ export const KeysListTab = ({ realmComponents }: KeysListTabProps) => {
             cellFormatters: [emptyFormatter()],
             transforms: [cellWidth(10)],
           },
+          {
+            name: "notAfter",
+            displayKey: "realm-settings:notAfter",
```
Comment (thomasdarimont) [inline js/apps/admin-ui/src/realm-settings/keys/KeysListTab.tsx:197]: Perhaps introduce a new translation key here "realm-settings:expiresAt" ?

File: core/src/main/java/org/keycloak/crypto/KeyWrapper.java [hunk 3]
```diff
@@ -195,4 +198,16 @@
...
         return key;
     }
+
+    public byte[] getSHA1Thumbprint() {
+        if (certificate == null) {
+            return null;
+        }
+        try {
+            // using SHA1 for certificate thumbprint should be fine
+            return MessageDigest.getInstance("SHA1").digest(certificate.getEncoded());
```
[the remark this answers is not in the document]
Comment (ahus1) [reply core/src/main/java/org/keycloak/crypto/KeyWrapper.java:208]: SHA-1 has been deprecated for some time, and shouldn't be used: https://en.wikipedia.org/wiki/SHA-1 around certificates (although this is "only" the fingerprint of the certificate itself, not the signed hash of the public key AFAIK).

I see that you manually create a hash for a certificate's byte stream. I wonder if there is a more ready-to-use Java mechanism to get a fingerprint of a certificate than this. Did you check that this matches for example the fingerprint provided by OpenSSL on the command line?

[the remark this answers is not in the document]
Comment (pedroigor) [reply core/src/main/java/org/keycloak/crypto/KeyWrapper.java:208]: That particular claim attribute is specific to SHA1. See https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.7 and https://datatracker.ietf.org/doc/html/rfc7517#page-9.

But yeah, SHA-1 should be deprecated in favor of SHA-256 as it is less likely to collisions. Libraries like Nimbus already deprecated using SHA-1 in thumbprints.

File: core/src/main/java/org/keycloak/jose/jws/JWSBuilder.java [hunk 4]
```diff
@@ -74,6 +76,7 @@ protected String encodeHeader(String sigAlgName) {
 
         if (type != null) builder.append(",\"typ\" : \"").append(type).append("\"");
         if (kid != null) builder.append(",\"kid\" : \"").append(kid).append("\"");
+        if (thumbprint != null) builder.append(",\"x5t\" : \"").append(thumbprint).append("\"");
```
Comment (pedroigor) [inline core/src/main/java/org/keycloak/jose/jws/JWSBuilder.java:79]: Is this related to exposing more info in admin UI?

If we are about to add support for adding the thumbprint to the JWS header, we should probably consider adding `x5t#S256`.

**Review by pedroigor — CHANGES_REQUESTED** [review]

**Comment by thomasdarimont** [comment]
I'm closing this PR in favor of #23025 which focusses on displaying the expiration date of a certificate in Keys List in admin-ui.

**Pull request closed without merging**